recitation_3/structs-unions.md

```diff
@@ -0,0 +1,281 @@
+    int x;
+    int y;
+};
+```
+
+This struct definition must be written at the top-level (e.g., it cannot be
+inside of a function body); don’t forget the trailing ’;’!
+
+The above struct definition creates a new type named "struct Point", which we
+can define and use like so:
+
+```c
+struct Point pt;
+
+pt.x = 2;
+pt.y = 3;
+```
+
+Structs are so-called because they describe the in-memory structure of a type.
+This is what `pt` looks like in memory:
+
+```
++---+---+---+---+---+---+---+---+
+|    .x = 2     |    .y = 3     |
++---+---+---+---+---+---+---+---+
+|----------- struct pt ---------|
+```
+
+### typedef
+
+A `typedef` is a directive that creates an alias for an existing type. We can use this directive to simplify the syntax of declaring instances of a struct like so:
+
+```c
+typedef struct {
+    int x;
+    int y;
+} Point;
+
+Point pt;
+
+pt.x = 2;
+pt.y = 3;
+```
+
+Notice how we no longer need to write `struct` in the type of `pt`. The above code has the same effect as the equivalent code for structs. The advantage of using typedef structs is that it makes the code cleaner and easier to read.
+
+An advatange of structs is that they don't need to have the same type of data; the above example could easily be expressed as an array since the struct only has members of type `int`. For example, the following
+struct MyString allows us to quickly access the length of a string without
+scanning it for a null terminator byte:
+
+```c
+struct MyString {
+    unsigned int len;
+    char *s = str;
+};
+ 
+struct MyString s;
+s.str = "abc";
+s.len = strlen("abc");
+```
+
+The size and layout of each struct is determined and fixed at compile time.
+Struct definitions may contains fields of any type, including pointers, arrays,
+and other structs (as long as they don’t contain themselves):
+
+```c
+struct TwoPoints {
+    struct Point p1;
+    struct Point p2;
+};
+
+struct Buffer {
+    unsigned int len;
+    char buf[1024];
+};
+
+struct Foo {
+    struct TwoPoints pp;
+    struct Foo f; // compiler error: a struct cannot contain itself
+};
+```
+
+Even though a struct cannot contain itself, it can contain a pointer to itself. This can allow for self-referential structs like so:
+
+```c
+struct IntListNode {
+    int data;
+    struct IntListNode *next;
+};
+```
+
+This can be helpful in creating more complex data structure such as linked lists -- a struct could contain a pointer to the next struct in its linked list.
+
+### Struct Padding
+
+In the `MyString` example above, the struct contains an `unsigned int` and a `char *`. We know that these types have different sizes -- an unsigned int is 4 bytes, while a pointer is always 8 bytes. 
+
+It may be intuitive to think that the total size of this struct is 4 + 8 = 12 bytes. However due to struct padding, the size of the struct is actually more than the sum of each element. 
+
+The layout would look like this:
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
```

Comment:
This isn't formatted right because it's too wide.. I'm not sure how to split it up though